File: app/src/main/java/br/com/menuvem/lojista/domain/model/TendenciaPreco.kt
package br.com.menuvem.lojista.domain.model

/** Tendência de preço de um insumo com base nas últimas 3 compras. */
enum class TendenciaPreco {
    ALTA,     // Preço subiu
    ESTAVEL,  // Preço manteve-se próximo
    QUEDA;    // Preço caiu

    companion object {
        /**
         * Calcula a tendência com base em uma lista de preços históricos
         * (do mais antigo para o mais recente).
         * Usa variação percentual entre o mais recente e o mais antigo da janela.
         */
        fun calcular(precos: List<Double>): TendenciaPreco {
            if (precos.size < 2) return ESTAVEL
            val janela = precos.takeLast(3)
            val primeiro = janela.first()
            val ultimo = janela.last()
            if (primeiro == 0.0) return ESTAVEL
            val variacao = (ultimo - primeiro) / primeiro
            return when {
                variacao > 0.02  -> ALTA     // > +2%
                variacao < -0.02 -> QUEDA    // < -2%
                else             -> ESTAVEL
            }
        }
    }
}

File: lojista/lib/domain/model/tendencia_preco.dart
/// Tendência de preço de um insumo com base nas últimas 3 compras.
enum TendenciaPreco {
  alta, // Preço subiu
  estavel, // Preço manteve-se próximo
  queda; // Preço caiu

  /// Calcula a tendência com base em uma lista de preços históricos (do mais
  /// antigo para o mais recente). Usa variação percentual entre o mais
  /// recente e o mais antigo da janela.
  static TendenciaPreco calcular(List<double> precos) {
    if (precos.length < 2) return TendenciaPreco.estavel;
    final janela = precos.length > 3
        ? precos.sublist(precos.length - 3)
        : precos;
    final primeiro = janela.first;
    final ultimo = janela.last;
    if (primeiro == 0.0) return TendenciaPreco.estavel;
    final variacao = (ultimo - primeiro) / primeiro;
    if (variacao > 0.02) return TendenciaPreco.alta;
    if (variacao < -0.02) return TendenciaPreco.queda;
    return TendenciaPreco.estavel;
  }
}
